### DISCUSSION TITLE
integrations/openvino/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8086] [Docs]
# integrations/openvino/

Discover the power of deploying your Ultralytics YOLOv8 model using OpenVINO format for up to 10x speedup vs PyTorch.

https://docs.ultralytics.com/integrations/openvino/

**Comment by YouROS12** [comment]
I wanna do inferences with my exported pre-trained yolov8 model using Raspberry PI with a Neural Network Computing 2 (NCS2) plugged in.. but its doesn't support opset 11 or above.. only opset10.. I specified onnx opset to 10 and convert it to Openvino but it didn't work. 

Anyway please to export openvino model directly in opset10 ? Does even yolov8 support this specified opset format ? 

Thank you.

**Reply by glenn-jocher** [reply]
@YouROS12 currently, YOLOv8 models are exported to ONNX with the latest opset version by default, which may not be compatible with certain versions of OpenVINO that support only up to opset 10. However, you can try exporting the YOLOv8 model to ONNX specifying opset version 10 and then convert it to OpenVINO format. Here's how you can do it:

1. Export the YOLOv8 model to ONNX with opset 10:
   ```python
   from ultralytics import YOLO

   # Load a pretrained YOLOv8 model
   model = YOLO('yolov8n.pt')

   # Export the model to ONNX with opset version 10
   model.export(format='onnx', opset=10)
   ```

2. Use the OpenVINO Model Optimizer to convert the ONNX model to OpenVINO's Intermediate Representation (IR) format. Make sure to specify the correct paths to your ONNX model and output directory:
   ```bash
   mo --input_model path/to/yolov8n.onnx --output_dir path/to/openvino_model --input_shape [1,3,640,640] --data_type FP16
   ```

3. After conversion, you should have the `.xml` and `.bin` files in your specified output directory. You can then use these files with the OpenVINO Inference Engine on your Raspberry Pi with NCS2.

Please note that while YOLOv8 supports exporting to ONNX with a specified opset, some operations used in YOLOv8 may not be supported in older opset versions, which could lead to errors during export or inference. If you encounter issues, it's recommended to check the compatibility of YOLOv8's operations with opset 10 or consider updating your OpenVINO toolkit to a version that supports the latest ONNX opset versions.

**Reply by YouROS12** [reply]
Where can I find documentation about YOLOv8's operations with opset 10 ? 
Updating OpenVINO toolkit is not an option because NCS2 works only at opset 10 and lower.

**Reply by pderrenger** [reply]
@YouROS12 the documentation for ONNX opset versions and the operations they support can be found on the ONNX GitHub repository or the official ONNX documentation. However, there isn't specific documentation for YOLOv8's operations with opset 10, as this would be a combination of YOLOv8's implementation details and ONNX's opset specifications.

To check which operations are supported in opset 10, you can refer to the ONNX [Operators documentation](https://github.com/onnx/onnx/blob/main/docs/Operators.md) for that specific version. This documentation lists all the operators available in opset 10 and their details.

For YOLOv8, you would need to look at the model's architecture and identify which operations are used. Then, cross-reference these operations with the ONNX opset 10 documentation to see if they are supported. If certain operations are not supported in opset 10, you might need to modify the model or the export script to use alternative operations that are compatible with opset 10.

Since updating OpenVINO is not an option for you, if you find that YOLOv8 uses operations not supported in opset 10, you might need to consider using a different model architecture that is compatible with opset 10, or reach out to the community or Ultralytics for further assistance on how to proceed with your specific use case and hardware constraints.

**Comment by kanishkasw** [comment]
Hi Im getting the following error when attempting to convert the model to openVino. Tried with both 2024.1.0 and 2024.0.0. Im on ubuntu 20.04 with python 3.8
$ yolo export model=runs/classify/train/weights/best.pt format=openvino
Ultralytics YOLOv8.2.31 🚀 Python-3.8.10 torch-2.3.1+cu121 CPU (Intel Core(TM) i9-7900X 3.30GHz)
YOLOv8n-cls summary (fused): 73 layers, 1438723 parameters, 0 gradients, 3.3 GFLOPs

PyTorch: starting from 'runs/classify/train/weights/best.pt' with input shape (1, 3, 320, 320) BCHW and output shape(s) (1, 3) (2.8 MB)
OpenVINO: export failure ❌ 0.0s: module 'openvino' has no attribute '__version__'
Traceback (most recent call last):
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/lib/python3.8/site-packages/ultralytics/cfg/__init__.py", line 591, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/lib/python3.8/site-packages/ultralytics/engine/model.py", line 602, in export
    return Exporter(overrides=args, _callbacks=self.callbacks)(model=self.model)
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/lib/python3.8/site-packages/ultralytics/engine/exporter.py", line 308, in __call__
    f[3], _ = self.export_openvino()
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/lib/python3.8/site-packages/ultralytics/engine/exporter.py", line 142, in outer_func
    raise e
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/lib/python3.8/site-packages/ultralytics/engine/exporter.py", line 137, in outer_func
    f, model = inner_func(*args, **kwargs)
  File "/home/Kanishka/v5.0/YOLOV8_classification/.venv/lib/python3.8/site-packages/ultralytics/engine/exporter.py", line 454, in export_openvino
    LOGGER.info(f"\n{prefix} starting export with openvino {ov.__version__}...")
AttributeError: module 'openvino' has no attribute '__version__'

**Reply by pderrenger** [reply]
@kanishkasw hi there,

Thank you for reaching out and providing detailed information about the issue you're encountering. It looks like you're experiencing an error when attempting to export your model to OpenVINO format. Let's work through this together.

First, let's ensure that you are using the latest versions of `torch` and `ultralytics`. Please run the following commands to update your packages:

```bash
pip install --upgrade torch ultralytics
```

Next, it seems the error is related to the OpenVINO package. The error message `module 'openvino' has no attribute '__version__'` suggests there might be an issue with the OpenVINO installation. Please make sure you have the correct version of OpenVINO installed. You can install it using the following command:

```bash
pip install openvino==2023.0.1
```

Once you've updated the packages, try exporting the model again:

```bash
yolo export model=runs/classify/train/weights/best.pt format=openvino
```

If the issue persists, could you please provide a minimum reproducible example? This will help us investigate the problem more effectively. You can find guidance on creating a minimum reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, here's a Python example to ensure everything is set up correctly:

```python
from ultralytics import YOLO

# Load a YOLOv8n PyTorch model
model = YOLO("runs/classify/train/weights/best.pt")

# Export the model to OpenVINO format
model.export(format="openvino")  # creates 'yolov8n_openvino_model/'

# Load the exported OpenVINO model
ov_model = YOLO("yolov8n_openvino_model/")

# Run inference
results = ov_model("https://ultralytics.com/images/bus.jpg")
```

For more detailed instructions on using OpenVINO, please refer to our [OpenVINO Integration Guide](https://docs.ultralytics.com/integrations/openvino/).

Feel free to reach out if you have any further questions or run into any other issues. We're here to help! 😊

**Comment by kimchi2020** [comment]
hey i am working on object detection in videos and came across openvino that helps in increasing speed. how can i use openvino and yolov8 together to find objects in a video

**Reply by glenn-jocher** [reply]
@kimchi2020 hi there! It's great to hear that you're exploring OpenVINO to enhance the speed of your object detection tasks in videos. OpenVINO can indeed provide significant performance improvements, especially on Intel hardware.

To get started with using YOLOv8 and OpenVINO together, follow these steps:

### Exporting YOLOv8 to OpenVINO Format

1. **Export the Model**:
   You can export your YOLOv8 model to the OpenVINO format using the following Python code:
   ```python
   from ultralytics import YOLO

   # Load a YOLOv8n PyTorch model
   model = YOLO("yolov8n.pt")

   # Export the model to OpenVINO format
   model.export(format="openvino")  # creates 'yolov8n_openvino_model/'
   ```

   Alternatively, you can use the CLI:
   ```bash
   yolo export model=yolov8n.pt format=openvino  # creates 'yolov8n_openvino_model/'
   ```

2. **Run Inference with the Exported Model**:
   After exporting, you can run inference using the exported OpenVINO model. Here’s how you can do it in Python:
   ```python
   from ultralytics import YOLO

   # Load the exported OpenVINO model
   ov_model = YOLO("yolov8n_openvino_model/")

   # Run inference
   results = ov_model("https://ultralytics.com/images/bus.jpg")
   ```

   Or using the CLI:
   ```bash
   yolo predict model=yolov8n_openvino_model source='https://ultralytics.com/images/bus.jpg'
   ```

### Running Inference on Videos

To process videos, you can modify the inference code to read frames from a video file and run the model on each frame. Here’s an example in Python:
```python
import cv2
from ultralytics import YOLO

# Load the exported OpenVINO model
model = YOLO("yolov8n_openvino_model/")

# Open the video file
video_path = "path/to/your/video.mp4"
cap = cv2.VideoCapture(video_path)

# Loop through the video frames
while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Run inference on the frame
    results = model(frame)

    # Visualize the results on the frame
    annotated_frame = results[0].plot()

    # Display the annotated frame
    cv2.imshow("YOLOv8 OpenVINO Inference", annotated_frame)

    # Break the loop if 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()
```

### Additional Resources

For more detailed information and advanced usage, you can refer to the [OpenVINO documentation](https://docs.ultralytics.com/integrations/openvino/). This guide covers everything from exporting models to running inference and optimizing performance.

Feel free to reach out if you have any more questions. Happy coding! 🚀

**Reply by kimchi2020** [reply]
Traceback (most recent call last):
  File "/home/kimchi/Desktop/yolov8/onnx.py", line 18, in <module>
    results = model(frame)
              ^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/model.py", line 174, in __call__
    return self.predict(source, stream, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/model.py", line 442, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/predictor.py", line 168, in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 35, in generator_context
    response = gen.send(None)
               ^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/predictor.py", line 234, in stream_inference
    self.model.warmup(imgsz=(1 if self.model.pt or self.model.triton else self.dataset.bs, 3, *self.imgsz))
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/nn/autobackend.py", line 626, in warmup
    import torchvision  # noqa (import here so torchvision import time not recorded in postprocess time)
    ^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torchvision/__init__.py", line 6, in <module>
    from torchvision import _meta_registrations, datasets, io, models, ops, transforms, utils
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torchvision/models/__init__.py", line 2, in <module>
    from .convnext import *
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torchvision/models/convnext.py", line 8, in <module>
    from ..ops.misc import Conv2dNormActivation, Permute
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torchvision/ops/__init__.py", line 1, in <module>
    from ._register_onnx_ops import _register_custom_op
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torchvision/ops/_register_onnx_ops.py", line 5, in <module>
    from torch.onnx import symbolic_opset11 as opset11
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torch/onnx/__init__.py", line 59, in <module>
    from ._internal.onnxruntime import (
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torch/onnx/_internal/onnxruntime.py", line 36, in <module>
    import onnx
  File "/home/kimchi/Desktop/yolov8/onnx.py", line 18, in <module>
    results = model(frame)
              ^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/model.py", line 174, in __call__
    return self.predict(source, stream, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/model.py", line 442, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/predictor.py", line 168, in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 35, in generator_context
    response = gen.send(None)
               ^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/engine/predictor.py", line 261, in stream_inference
    self.results = self.postprocess(preds, im, im0s)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/models/yolo/detect/predict.py", line 25, in postprocess
    preds = ops.non_max_suppression(
            ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kimchi/Desktop/yolov8/myvenv/lib/python3.12/site-packages/ultralytics/utils/ops.py", line 291, in non_max_suppression
    i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        ^^^^^^^^^^^^^^^
AttributeError: partially initialized module 'torchvision' has no attribute 'ops' (most likely due to a circular import). Did you mean: 'os'?

I am getting this error, how to get rid of it and what does it mean

**Reply by glenn-jocher** [reply]
Hi @kimchi2020,

Thank you for reaching out! The error you're encountering seems to be related to a circular import issue with the `torchvision` library. This can happen if there's a naming conflict or an import loop in your code.

### Steps to Resolve the Issue:

1. **Check for Naming Conflicts**:
   Ensure that your script name (`onnx.py`) is not conflicting with any library names. Rename your script to something unique, like `run_inference.py`.

2. **Verify Installation**:
   Make sure you have the latest versions of `torch`, `torchvision`, and `ultralytics` installed. You can update them using:
   ```bash
   pip install --upgrade torch torchvision ultralytics
   ```

3. **Reproduce the Issue**:
   If the issue persists, please provide a minimum reproducible example. This helps us diagnose the problem more effectively. You can refer to our guide on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

### Example Code for Inference with OpenVINO:

Here's a quick example of how to export your YOLOv8 model to OpenVINO and run inference on a video:

```python
import cv2
from ultralytics import YOLO

# Load a YOLOv8n PyTorch model
model = YOLO("yolov8n.pt")

# Export the model to OpenVINO format
model.export(format="openvino")  # creates 'yolov8n_openvino_model/'

# Load the exported OpenVINO model
ov_model = YOLO("yolov8n_openvino_model/")

# Open the video file
video_path = "path/to/your/video.mp4"
cap = cv2.VideoCapture(video_path)

# Loop through the video frames
while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Run inference on the frame
    results = ov_model(frame)

    # Visualize the results on the frame
    annotated_frame = results[0].plot()

    # Display the annotated frame
    cv2.imshow("YOLOv8 OpenVINO Inference", annotated_frame)

    # Break the loop if 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release the video capture object and close the display window
cap.release()
cv2.destroyAllWindows()
```

For more detailed information on exporting and using OpenVINO, please refer to our [OpenVINO Integration Guide](https://docs.ultralytics.com/integrations/openvino/).

Feel free to reach out if you have any more questions or need further assistance. Happy coding! 🚀

**Reply by YA12SHYAM** [reply]
Hi @glenn-jocher , what if i want to use intel GPU's  for the inference, using device=GPU doesn't seems to works , it looks for cuda i.e nvidia gpu

**Reply by glenn-jocher** [reply]
To use Intel GPUs for inference, export your YOLOv8 model to OpenVINO format. This allows you to leverage Intel hardware, including GPUs, for accelerated inference. You can export using the command `model.export(format="openvino")` and then run inference with the exported model. For more details, please refer to the OpenVINO documentation.

**Reply by YA12SHYAM** [reply]
ok, just to confirm i can't use ultralytics python api for inference with intel gpu like i usually do within the ultralytics ecosystem  for example ?
```
from ultralytics import YOLO
model =  YOLO(path/to/model)
```

here the path/to/model works with onnx, engine, opnevino;s xml , but it use cpu in case of openvino , i want to use my intel GPU

**Reply by glenn-jocher** [reply]
@YA12SHYAM to use Intel GPUs with YOLOv8, export your model to OpenVINO format and utilize the OpenVINO Runtime, which supports Intel GPUs. The standard Ultralytics Python API defaults to CPU for OpenVINO models. For detailed steps, refer to the OpenVINO documentation.

**Comment by Ranjith3436** [comment]
You mentioned that the core CPU inference model takes 23.53 ms. I would like to know if the processing time decreases when we increase the number of CPU cores ?  currently i am using 12 thread CPU.

**Reply by glenn-jocher** [reply]
@Ranjith3436 increasing the number of CPU cores can potentially reduce processing time, as it allows for parallel processing, but the actual speedup will depend on the model's ability to leverage multi-threading effectively.

**Reply by Ranjith3436** [reply]
how do i can know about model's ability?

**Reply by glenn-jocher** [reply]
To assess a model's ability to utilize multiple CPU cores, you can check its support for multi-threading or parallel processing in the documentation or by running benchmarks on your hardware configuration.

**Comment by EduardoNew** [comment]
I encountered some difficulties while trying to run the official video demo. Here is my code:

from ultralytics import YOLO

def main():
    # Load a YOLOv8n PyTorch model
    model = YOLO("yolov8n.pt")

    # Export the model
    model.export(format="openvino")  # creates 'yolov8n_openvino_model/'

    # Load the exported OpenVINO model with explicit task definition
    ov_model = YOLO("yolov8n_openvino_model\\", task='detect')  # Explicitly define task as 'detect'

    # Run inference
    results_original = model("D:\\python_YOLO_test\\bus.jpg")

    # Run ov inference
    results_ov = ov_model("D:\\python_YOLO_test\\bus.jpg")


if __name__ == "__main__":
    main()
Additional notes: My environment is openvino==2024.0.0, and I executed pip install --upgrade torch ultralytics.

The issue I encountered is as follows:

Results saved to D:\python_YOLO_test
Ultralytics YOLOv8.2.57 🚀 Python-3.11.9 torch-2.1.1+cu118 CPU (11th Gen Intel Core(TM) i7-11800H 2.30GHz)
YOLOv8n summary (fused): 168 layers, 3,151,904 parameters, 0 gradients, 8.7 GFLOPs

PyTorch: starting from 'yolov8n.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (6.2 MB)

OpenVINO: starting export with openvino 2024.0.0-14509-34caeefd078-releases/2024/0...
OpenVINO: export success ✅ 2.8s, saved as 'yolov8n_openvino_model\' (12.3 MB)

Export complete (5.2s)
Results saved to D:\python_YOLO_test
Predict:         yolo predict task=detect model=yolov8n_openvino_model imgsz=640  
Validate:        yolo val task=detect model=yolov8n_openvino_model imgsz=640 data=coco.yaml  
Visualize:       https://netron.app

image 1/1 D:\python_YOLO_test\bus.jpg: 640x480 4 persons, 1 bus, 1 stop sign, 324.0ms
Speed: 24.0ms preprocess, 324.0ms inference, 21.0ms postprocess per image at shape (1, 3, 640, 480)
Loading yolov8n_openvino_model for OpenVINO inference...
Using OpenVINO LATENCY mode for batch=1 inference...

Process finished with exit code -1073741819 (0xC0000005)

Process finished with exit code -1073740791 (0xC0000409)
The model was successfully exported but did not perform inference, instead, it terminated directly. How should I resolve this issue?

**Reply by glenn-jocher** [reply]
@EduardoNew the exit code 0xC0000005 often indicates a memory access violation. Let's address this step-by-step:

1. **Path Format**: Try using forward slashes in your OpenVINO model path: `"yolov8n_openvino_model/"` instead of backslashes.

2. **OpenVINO Version**: Our benchmarks show best results with OpenVINO 2023.0.1. Try downgrading:
```bash
pip install openvino==2023.0.1
```

3. **Hardware Compatibility**: Ensure you have the latest Intel drivers installed from [Intel's official site](https://www.intel.com/content/www/us/en/download-center/home.html).

If issues persist, please create a Minimum Reproducible Example and open a GitHub Issue following our bug report guidelines. We'll be happy to investigate further!

**Comment by Prabindas001** [comment]


**Reply by allenhack638** [reply]
Facing a similar issue, it would be great if ultralytics can allow passing custom configuration parameters or load from env into the class AutoBackend(nn.Module) --> def _init_() for openVino models.

**Reply by glenn-jocher** [reply]
@Prabindas001 could you share the exact command/code you ran, the full traceback as text, and diagnostics from `yolo checks` (or `!yolo checks` in notebooks)? This will help us reproduce and debug.

@allenhack638 thanks for the suggestion—today the API exposes device selection for OpenVINO models (e.g., `device="intel:gpu"`), but deeper OpenVINO compile properties aren’t surfaced; for now set them via the native runtime as outlined in our [OpenVINO inference optimization guide](https://docs.ultralytics.com/guides/optimizing-openvino-latency-vs-throughput-modes/). If you’d like this upstream, please open a focused feature request or PR proposing an API like the example below; keep changes minimal, add tests/docs, and avoid edits to `default.yaml`—we’ll review but can’t promise timelines.

```python
# proposed shape (not implemented yet)
results = model.predict(
    source="video.mp4",
    device="intel:gpu",
    ov_config={"PERFORMANCE_HINT": "THROUGHPUT", "NUM_STREAMS": "AUTO"}
)
```
